### ISSUE TITLE
Add postal cities (for USA)

### ISSUE THREAD

**Issue by nvkelso** [whosonfirst-data/whosonfirst-data#202]
It would be helpful to know what the USPS thinks the "city" is when you're not in a city. This can be derived from the ZCTA from US Census and looking at the locality name.

From: https://github.com/pelias/pelias/issues/167#issuecomment-178084524

Right now we return:

`110A Sage Street, Westwood, Davis, CA`

Where Westwood is a neighbourhood just outside of the incorporated city of Davis. 

When it's more conventional to provide a mailing label like:

`110A Sage Street, Davis, CA`

There's probably related logic necessary to recommend that the neighbourhood be included in the label or not.

**Comment by stepps00** [comment]
Dupe of #72

**Comment by nvkelso** [comment]
(Not a dup of #72.)

**Comment by nvkelso** [comment]
Another example:

`4141 Riverport Dr, St Louis` is technically the city is Maryland Heights. But many people would expect to allow St Louis as the city name as that's what USPS uses for postal delivery.

**Comment by thisisaaronland** [comment]
We should just add a `wof:postal_city` (or something like `postal_locality` to be consistent) property to the properties hash, where appropriate. Something like what we've talked about doing to for "airport(s) that service this place".

**Comment by nvkelso** [comment]
So on postcode features add `wof:postcode_locality`?

Since postcodes (at least in the USA) are based on mail delivery areas that don't line up with other placetype geometries I can see adding property to postcode features (versus new placetype and geoms).

**Comment by nvkelso** [comment]
@orangejulius @dianashk Do you import postcode features now for PIPing? We're proposing adding `wof:postcode_locality` property to existing postcode features (rather than inventing a new placetype). Would that work for you? The benefit would be a single PIP instead of needing to do a double PIP.

**Comment by nvkelso** [comment]
(Often USPS prefers a single locality name, but allows several other alt names for a zipcode, loosely related to actual localities in that postal delivery zone. So we may want `wof:postal_locality` and an optional list of variant names that are also "legal" for that place?)

**Comment by thisisaaronland** [comment]
Yeah – if the property is part of a `postalcode` record then perhaps it makes sense to default to a list of locality IDs, rather than a singleton. It seems inevitable there will be instances where a postalcode has multiple "postal cities".

**Comment by dianashk** [comment]
@nvkelso, no we don't use postcodes for the pip service. The sheer volume of postcodes currently overloads our pip service as it holds all the data in memory. If the number of postcodes that have this new property is < all postcodes in the world then we might be able to start using just that subset.

**Comment by nvkelso** [comment]
It would be USA only at first.

**Comment by orangejulius** [comment]
@nvkelso this is awesome. How many postal codes currently have this mapping? We have some ideas for a PIP service rearchitecture that would allow us to store a bit more data, so it might be possible to fit.

**Comment by nvkelso** [comment]
We can harvest this information from the USPS:

- https://tools.usps.com/go/ZipLookupResultsAction!input?postalCode=80241&state=Select&address2=&zip=&address1=&resultMode=2&city=&urbanCode=&companyName=

It has preferred locality, allowed alternate localities, and prohibited localities (!?). 

It's unclear if we could harvest _all_ the data from the USPS web tool. There are some alternatives (purchase):

- USPS: https://ribbs.usps.gov/index.cfm?page=address_info_systems
- 3rd party like https://www.zipinfo.com/products/z5/z5.htm
    - About 71,000 records (42,300 "preferred" records plus 28,700 "alias" records)

I further propose importing into WOF as:

- `wof:postal_locality` - a single locality id
    - often the locality where the postal distribution center is located / postmaster lives
- `wof:postal_locality_alt` - a list of locality ids
    - where USPS recognizes as valid alternative localities (often a zipcode will cover multiple localities), does not include the primary wof:postal_locality id.
- `wof:postal_locality_funky` - a list of locality ids
    - where USPS says mail may get lost because of confusing address and asks for the locality name to NOT be used

This assumes we can turn each USPS locality name (like `San Francisco CA`) into a WOF locality ID (which is probably true. It also depends on how much WOF consumers like Pelias would like just a text value versus a WOF id pointer. 

@orangejulius @dianashk would you be okay with WOF id's?

**Comment by nvkelso** [comment]
This should probably be implemented on the postcode features themselves as attributes instead of a new placetype (with duplicate / weird polygons).

Related Pelias work in that direction is here: https://github.com/pelias/wof-admin-lookup/pull/218,

**Comment by thisisaaronland** [comment]
What @nvkelso said.

**Comment by stepps00** [comment]
Bumping this issue.. I'm working on creating a postalcode <> locality name mapping for the US right now. It'll likely include:

- postalcode (ex: 97220)
- preferred name (ex: Portland)
- alternate names (ex: Maywood Park)

I'd like to complete this mapping for all countries, but I'll start with the US. [This comment](https://github.com/whosonfirst-data/whosonfirst-data/issues/202#issuecomment-315830435) mentions using `wof:id` values to populate new properties, but I'm working with locality names... so I'll need to figure out how to actually import the mapping intto WOF once finalize the matches.

cc @missinglink @orangejulius

**Comment by orangejulius** [comment]
@stepps00 really cool. I'd love to hear more about how you plan to generate this.

Since there was last movement on this particular Github issue, we've made the https://github.com/pelias/lastline project which has made good progress in solving this problem for Pelias.

Essentially, it uses looks at all the records in OSM with both a city name and postal code, and uses that to create a mapping between the two (as well as WOF ID by finding the nearest locality with the same city name). Because we can count the number of occurrences of each mapping, it's fairly resilient to data errors (although there certainly are some), and also provides alternate names by ranking the mappings by count.

**Comment by missinglink** [comment]
@stepps00 check out https://github.com/pelias/wof-admin-lookup/blob/master/src/data/USA.tsv

**Comment by vicchi** [comment]
@stepps00 Bumping on your April 16th bump ... did you make any progress on or work out what a WOF postcode place would look like for this? If so, would be good to use as a starting reference point for UK postal towns, as being discussed over at #1871

**Comment by stepps00** [comment]
@vicchi this is still in progress, but something I should pick back up and merge into WOF.. I'll craft a PR in the `whosonfirst-data-admin-us` repo that gives an example of the property proposal and flag you.

I also owe a comment or two in #1871 :eyes:

**Comment by stepps00** [comment]
> @stepps00 check out https://github.com/pelias/wof-admin-lookup/blob/master/src/data/USA.tsv

@missinglink Thanks for this link.. though I may have missed what the columns represent in this file. I've made the assumption that the last column is the amount of times that locality name was found per postalcode, is that true?

> @stepps00 really cool. I'd love to hear more about how you plan to generate this.

My first run had quite a few questionable locality names for postal codes in the US, so I took another shot at generating this postal code <> locality mapping using two sources - Pelias' [USA.tsv](https://github.com/pelias/wof-admin-lookup/blob/master/src/data/USA.tsv) lookup table, and WOF's venue data. I may have a third source, though I need to verify the license information.

Note that I'm using 5-char (`XXXXX`) ZIPs, rather than ZIP+4 codes (`XXXXX-XXXX`).

Export of the run against Pelias' lookup table: [here](https://gist.github.com/stepps00/d0f2f2561a85d12ca12c825279b6bbac).
Export of WOF's venue data: [here](https://gist.github.com/stepps00/0d5a89e172df86bc1c0d9724a0d80453).

With the Pelias file, I converted ZIP+4 codes to 5-char ZIP codes, cleaned up some non-US and erroneous codes, then created a list of key (locality name), value (amount of times that name occurs) pairs for each ZIP. Then, sorted the list by occurrences. 

With WOF, I crawled each venue repo in the US to pull out `'sg:city'` and `'sg:postcode'` in each record, then create the same type of postal code <> locality mapping.

Examples:

```
10003	[('New York', 7482), ('Manhattan', 193), ('New York City', 12), ('Brooklyn', 7), ('New York ', 3), ('Nyc', 1)]
18509	[('Scranton', 50), ('Dunmore', 4)]
99503	[('Anchorage', 41)]
```

In these cases, the first name listed would be the "preferred" name for the postalcode, the others being "alternate" names. I still need to scrub some of the locality names ("Nyc" probably isn't valid) and combine these lists for import, but hoping to wrap this up soon.

**Comment by missinglink** [comment]
> I may have missed what the columns represent in this file

The columns are documented here:
https://github.com/pelias/wof-admin-lookup#contributing

**Comment by nvkelso** [comment]
The files look fairly consistent between Pelias and WOF venue runs, nice! I spot checked a few zipcodes I've lived or worked.

> In these cases, the first name listed would be the "preferred" name for the postalcode, the others being "alternate" names. I still need to scrub some of the locality names ("Nyc" probably isn't valid) and combine these lists for import, but hoping to wrap this up soon.

Agree we should strip alternate names when there is just a couple probably bogus instances. Perhaps would require at least X count (like more than 2 instances) or a % (like 1%) to be included as an alternate, as I've seen some rows where there are only 10 total count and some rows with thousands of total count. 

> 95476	[('Sonoma', 298), ('Eldridge', 1), ('Oakland', 1), ('Schellville', 1)]

I'd think we'd only have Sonoma as the default, and no alternates listed.

> 95536	[('Petrolia', 21), ('Ferndale', 5)]

Probably keep Ferndale as an altnerate.

> 20781	[('Hyattsville', 2665), ('Edmonston', 1)]

I'd think we'd only have Hyattsville as the default, and no alternates listed.

> 20782	[('Hyattsville', 3723), ('University Park', 928), ('Park', 2)]

I'd think we'd only have Hyattsville as the default, with University Park as alternate (not including separate Park)

> 20706	[('Lanham', 9596), ('Glenn Dale', 15), ('Lanham-Seabrook', 7), ('Glenarden', 5), ('Greenbelt', 1)]

Where do you draw the line in this one? No Greenbelt, but what about the others?

**Comment by missinglink** [comment]
We have some test cases in https://github.com/pelias/wof-admin-lookup/blob/master/test/lookupStreamPostalCityTest.js which have been contributed when fixing postal city lookup bugs, these might be helpful for sanity checks.

**Comment by stepps00** [comment]
> The columns are documented here:
https://github.com/pelias/wof-admin-lookup#contributing

> We have some test cases in https://github.com/pelias/wof-admin-lookup/blob/master/test/lookupStreamPostalCityTest.js which have been contributed when fixing postal city lookup bugs, these might be helpful for sanity checks.

Perfect, these are both very helpful.

> Agree we should strip alternate names when there is just a couple probably bogus instances. Perhaps would require at least X count (like more than 2 instances) or a % (like 1%) to be included as an alternate, as I've seen some rows where there are only 10 total count and some rows with thousands of total count.

I've reworked the data a bit and created two new files:

- [This gist file](https://gist.github.com/stepps00/2e1ed01be91f3fd371eacf58f7fed134) is a combination of the runs against WOF and Pelias data, ordered by name occurrence per postalcode.
- And [this gist file](https://gist.github.com/stepps00/9cf0653c96464967b588938c7151f169) is a combination of the runs against WOF and Pelias data, ordered by name occurrence per postalcode.. but removes names that only occurred one or two times. 

Overall, this is not perfect, but worked pretty well. There are some postalcodes that result in no locality names, but in those cases, neither WOF or Pelias have strong locality name data. I've also spot-checked places I'm familiar with and the locality names look appropriate.

To get this into WOF, I'll propose:

- Running these combined, scrubbed results against the Pelias test cases
- We use a `mz:postal_locality` property to import these names into WOF postalcode records
  - The top ranked (first listed) locality name per postalcode would live in the `wof:postal_locality` property
  - The remaining locality names would live in the `wof:postal_locality_alt` property
  - In a handful of cases, we'd curate a `mz:postal_locality_funky` property, but this would mostly be an exercise to prove what;s possible for future work
- Holding off on any new placetype creation for now (something discussed in https://github.com/whosonfirst-data/whosonfirst-data-postalcode-us/pull/5.

cc @nvkelso

**Comment by missinglink** [comment]
Happy NY!

Did you guys come to some decision about postcodes codes which don't actually refer to a `locality` per-se?
I'm still a little concerned about the use of the terminology `wof:postal_locality` for those.

> **Gotchas**
> The term "Postal **City**" is a misnomer, some examples:
> 
> * [ZIP 11201](https://www.unitedstateszipcodes.org/11201/) corresponds to the "City" of "Brooklyn" (ie. a `borough` in WOF)
> * [ZIP 11374](https://www.unitedstateszipcodes.org/11374/) corresponds to the "City" of "Rego Park" (ie. a `neighbourhood` in WOF)
> * Some `campus/venue` placetypes: White House (ZIP 20500), LaGuardia Airport (ZIP 11371), Pearl Harbor, HI (ZIP 96860)

https://github.com/whosonfirst-data/whosonfirst-data-postalcode-us/pull/5#issuecomment-675361316

**Comment by nvkelso** [comment]
> Did you guys come to some decision about postcodes which don't actually refer to a locality per-se?

We deferred supporting a new **"postal locality"** placetype to a future date. 

Your examples show the full range of beautiful geo soup on offer, bravo! They are indeed a mix of placetypes / though not exactly the same as those features. The world is wild; our goal is to fill in the fake "city name" in a recognizable way useful for address validation. I think this proposal is a big step forward, though not a 100% solution.

In the meantime...

> I'm still a little concerned about the use of the terminology wof:postal_locality for those.

Per @thisisaaronland's comment in https://github.com/whosonfirst-data/whosonfirst-data-postalcode-us/pull/5#issuecomment-678369087, we settled on the `mz` namespace for storing these string values. 

Which changes @stepps00 comment (fixes typos?), to:

We will use a `mz:postal_locality` property to import these **names** into WOF postalcode records:
  - `mz:postal_locality` - _string property_ - stores the top ranked (first listed) locality name per postalcode would live in the 
  - `mz:postal_locality_alt` - _list of strings_ - any remaining locality names (optional)
  - `mz:postal_locality_funky` - _list of strings_ - In rare cases, we'd curate a  property. With the initial PR this would just be handful of places proving the data model.

**Comment by stepps00** [comment]
> Which changes @stepps00 comment (fixes typos?), to:

Ah, my mistake. Your comment with the `mz:` prefixes is correct (I've tweaked my comment).

Also, an update on this: out of an abundance of caution, I'm going to import _just_ locality names I've obtained from crawling the WOF venue data. It was pointed out that the Pelias lookup file contains names sourced from OSM, so I'd like to import the names from WOF's own venues as a first pass.

The resulting names being imported likely won't be impacted much, as I was able to harvest significantly more locality names from WOF. The Pelias file did have some unique names that weren't in WOF, but was mostly used to confirm what I'd pulled from WOF venues.

So, the next step is to actually craft a PR to get this imported. I opened https://github.com/whosonfirst-data/whosonfirst-data-postalcode-us/pull/5 mostly as a proof of concept.. so I'll likely revert the changes to the four records in that PR, then update all postalcode records in one go.. I expect to have this wrapped up by early next week.
